server/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -167,7 +167,6 @@ public RecoveryRef getRecoverySafe(long id, ShardId shardId) {
             throw new IndexShardClosedException(shardId);
         }
         assert recoveryRef.target().shardId().equals(shardId);
-        assert recoveryRef.target().indexShard().routingEntry().isPromotableToPrimary();
```

Comment:
Looks like this was added [here](https://github.com/elastic/elasticsearch/pull/93210/files#diff-54eba0f7c8583f1632e9b21c788d910cd5732df56df6a5f50010b246307cc0ceR167), I am also not sure I understand why, perhaps @kingherc remember and can confirm that the assertion is not significant?

Reply:
Not out of the top of my head. But going back to the code, I see we've made a special branch in `PeerRecoveryTargetService#doRecovery()` with `if (indexShard.routingEntry().isPromotableToPrimary() == false) {` for unpromotables that basically quick skips all recovery stages, and closes the RecoveryRef as well. So the point of the assertion at the time was that there should be no other coordination needed for unpromotables to justify getting the RecoveryRef.

Seeing though that now this PR introduces some sort of coordination between unpromotables, it probably makes to remove the assertion.

Reply:
(I did not fully review this PR, but feel free to tell me if I should)